### ISSUE TITLE
New freight for a new docker tag is not created if freight with a different tag but the same digest exists.

### ISSUE THREAD

**Issue by daftping** [akuity/kargo#1324]
# Checklist

* [X] I've searched the issue queue to verify this is not a duplicate bug report.
* [X] I've included steps to reproduce the bug.
* [X] I've pasted the output of `kargo version`.
* [ ] I've pasted logs, if applicable.

# Description
New freight for a new docker tag is not created if freight with a different tag but the same digest exists.

# Screenshots

# Steps to Reproduce
Create Warehouse
```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Warehouse
...
spec:
  subscriptions:
  - image:
      repoURL: daftping/demo
      semverConstraint: ^1
      tagSelectionStrategy: SemVer
status:
  observedGeneration: 1
```
Prepare an image 
```shell
docker tag nginx:latest daftping/demo:1.0.1
docker push daftping/demo:1.0.1
```

Freight is generated
```yaml
apiVersion: kargo.akuity.io/v1alpha1
id: c1f459dc7d8bc3cd649560c0d5b21d150d12e6eb
images:
- digest: sha256:4669f6671aca20a34c3dfcd017e84fb3cae40788ea664866eaea698e3dfe241c
  repoURL: daftping/demo
  tag: 1.0.1
kind: Freight
metadata:
  creationTimestamp: "2023-12-22T03:25:21Z"
  generation: 1
  labels:
    kargo.akuity.io/alias: wishing-rattlesnake
  name: c1f459dc7d8bc3cd649560c0d5b21d150d12e6eb
  namespace: kargo-demo
  ownerReferences:
  - apiVersion: kargo.akuity.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: Warehouse
    name: kargo-demo-daftping
    uid: 54581ba9-0f87-41c7-9185-2f6040313afd
  resourceVersion: "387982"
  uid: b5cdfe71-d25f-4818-b40a-c6305830d7eb
status:
  verifiedIn:
    prod: {}
    test: {}
    uat: {}
```
Re-tag and push a new image
```shell
docker tag daftping/demo:1.0.1 daftping/demo:1.0.2
docker push daftping/demo:1.0.2
```

**New freight is not created**

Prepare an image with different digest
```shell
docker tag nginx:1.24.0 daftping/demo:1.0.3
docker push daftping/demo:1.0.3
```

New freight successfully created
```yaml
apiVersion: kargo.akuity.io/v1alpha1
id: 9b15ff94d16d619c79514322798bdacbfebfdaa3
images:
- digest: sha256:1e0f779773b10a4e36f4aaa304afe4a6334f502a317542c45d58ab2e9c563f59
  repoURL: daftping/demo
  tag: 1.0.3
kind: Freight
metadata:
  creationTimestamp: "2023-12-22T17:48:32Z"
  generation: 1
  labels:
    kargo.akuity.io/alias: quoting-chicken
  name: 9b15ff94d16d619c79514322798bdacbfebfdaa3
  namespace: kargo-demo
  ownerReferences:
  - apiVersion: kargo.akuity.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: Warehouse
    name: kargo-demo-daftping
    uid: 54581ba9-0f87-41c7-9185-2f6040313afd
  resourceVersion: "489129"
  uid: 7732dbf8-d025-4ea1-988b-b528073ea98a
```

# Version
```
Client Version: v0.2.1
Server Version: v0.3.0-rc.3
```

# Logs

```
Paste any relevant application logs here.
```

**Comment by krancour** [comment]
@daftping nice find.

We just recently started using image digests instead of tags in calculating Freight IDs because of the opposite scenario...

Not that I condone the use of mutable tags like `latest`, but suppose someone wanted to subscribe to `latest` using the `NewestBuild` selection strategy. Using the digest instead of the tag allows for a new piece of Freight every time a new `latest` is pushed.

Clearly, I overlooked your scenario where you're re-tagging an existing image.

Hindsight being 20/20, it's clear that both the tag _and_ digest should be factored into calculating Freight IDs.

Thanks again for finding this.
